### PR TITLE
Enforce API key type on authentication

### PR DESCRIPTION
**Pull request by ywangd**
The PR adds enforcement for API key type at authentication time. Concretely, new cross-cluster API keys (#95714) can only be used on the dedicated remote cluster interface and the existing (rest) API key must not be used for new remote cluster communication. To make cross-cluster API keys actually usable after authentication, the PR also adds support for resolving their roles.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyAuthenticator.java [hunk 1]
```diff
@@ -57,6 +60,7 @@ public void authenticate(Context context, ActionListener<AuthenticationResult<Au
                     ? authResult.getException()
                     : Exceptions.authenticationError(authResult.getMessage());
                 logger.debug(() -> "API key service terminated authentication for request [" + context.getRequest() + "]", e);
+                // TODO: emit audit event
                 listener.onFailure(e);
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyAuthenticator.java:64]: We have an open issue (https://github.com/elastic/elasticsearch/issues/52551) for missing audit logs when ApiKeyService fails to authenticate. It impacts here as well, i.e. when API key type does not match, the authentication terminates, but no audit event is generated. I will fix it separately when working on the plan item for audit logging. More specifically, I plan to emit an authenticationFailure event here when the authentication "terminates". But not when it "continues", which shoud be fixed separately.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java [hunk 2]
```diff
@@ -994,6 +995,17 @@ void validateApiKeyCredentials(
             listener.onResponse(
                 AuthenticationResult.unsuccessful("document [" + docId + "] is [" + apiKeyDoc.docType + "] not an api key", null)
             );
+        } else if (apiKeyDoc.type != credentials.expectedType) {
+            listener.onResponse(
+                AuthenticationResult.terminate(
+                    Strings.format(
+                        "authentication expected API key type of [%s], but API key [%s] has type [%s]",
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:1002]: Since we have found a way to have the type information carried by `ApiKeyCredentials` instead of `Authenticator#Context`, I think it is now better to perform the type check in `ApiKeyService` instead of `ApiKeyAuthenticator`. A major advantage is that the type check can happen before we validate the API key secret. It's both faster and more secure.

File: x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/AbstractRemoteClusterSecurityWithMultipleRemotesRestIT.java [hunk 3]
```diff
@@ -67,7 +67,8 @@ public void testCrossClusterSearch() throws Exception {
                 { "index": { "_index": "cluster1_index1" } }
                 { "name": "doc1" }
                 { "index": { "_index": "cluster1_index2" } }
-                { "name": "doc2" }\n"""));
+                { "name": "doc2" }
+                """));
```
Comment (ywangd) [inline x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/AbstractRemoteClusterSecurityWithMultipleRemotesRestIT.java:71]: Some nit fixes.

File: x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityRestIT.java [hunk 4]
```diff
@@ -327,6 +344,25 @@ public void testNodesInfo() throws IOException {
...
+                    .put("cluster.remote.wrong_api_key_type.mode", "proxy")
+                    .put("cluster.remote.wrong_api_key_type.proxy_address", fulfillingCluster.getRemoteClusterServerEndpoint(0))
+                    .build()
+        );
+        final ResponseException e = expectThrows(
+            ResponseException.class,
+            () -> performRequestWithRemoteSearchUser(new Request("GET", "/wrong_api_key_type:*/_search"))
+        );
+        assertThat(e.getResponse().getStatusLine().getStatusCode(), equalTo(401));
+        assertThat(e.getMessage(), containsString("unable to authenticate user "));
```
Comment (n1v0lg) [inline x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityRestIT.java:363]: Not for this PR but currently, and really more a musing than an actionable suggestion. 

Currently, if a REST API key is configured (in place of a cross-cluster one) this is actually not that easy to detect, for a QC-side admin. 

All they get are generic 401 messages on CCS (and other remote requests), without the extra context that this is due to the wrong API key type (I think).

This isn't something we introduced with this change (same problem exists for e.g., key not found) and may be kind of tricky to address. So perhaps the best way to deal with this is just having a trouble-shooting section in docs that emphasizes common causes for a 401 here... 

Just jotting this down. As I said, not something to worry about now.

**Review by n1v0lg — APPROVED** [review]
LGTM! This came with a v neat and small change set. 

I'm pondering the following: before this new restriction, QC-side admins could use routes like `GET /_security/_authenticate` to get information about the API key they've been given, which could be useful for debugging. This is not possible anymore. I think that's fine, and there was never a requirement for QC admins to be able to do this. Just something that occurred to me. It's an exception to other kinds of credentials where it's possible to run `_authenticate` and friends. 

Also, optional but I don't think we covered secondary authentication tests (sorry if did and I missed it!).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java [hunk 5]
```diff
@@ -265,7 +270,25 @@ private RoleReferenceIntersection buildRoleReferencesForApiKey() {
         );
     }
 
+    // Package private for testing
+    RoleReferenceIntersection buildRoleReferencesForCrossClusterApiKey() {
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java:274]: Indeed. I was too busy at copy & paste ... Thanks!

File: x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityRestIT.java (hunk 4 above)
Comment (ywangd) [reply x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityRestIT.java:363]: Errr this turns out to be a test bug. Thanks for cathing it!

With #94998, we should report underlying error in the response. The real issue here is that the `remote_search` user was not created on the querying cluster. It was a native user created in the beginning of `testCrossClusterSearch` method and I didn't copy it in the new method. So It was actually reporting 401 of the `remote_search` on the querying cluster. I fixed it by merging it back to `testCrossClusterSearch`. Now the error response contains message about mismatched API key type. Also these errors are logging on the QC side so that QC admins will be able to see them right after configuring the remote cluster because the handshake would fail for the same error. In summary, I think the error should be pretty visible on the QC side.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:1002]: This is a good point. In theory we should not reveal any information unless authentication is successful. I was thinking to limit cross-cluster API keys as much as possible on the REST interface, i.e. not letting them to trigger hash computation. But that is probably not really necessary since it is probably easier for someone to get a valid API key ID and trigger hash compuation compared to getting a API key ID specific for cross-cluster usage.

Though I did consider the type information to be public, I did suggest prefix the ID with it. However since we are not revealing it at this stage, it is probably still overall better to place the check behind hash verification. It means extra milliseconds needs to be spent on first authentication. But it is cached afterwards and should have very minor overhead.

In summary, I moved the check after hash verification.

**Comment by ywangd** [comment]
> before this new restriction, QC-side admins could use routes like GET /_security/_authenticate to get information about the API key they've been given, which could be useful for debugging. This is not possible anymore. I think that's fine, and there was never a requirement for QC admins to be able to do this. Just something that occurred to me. It's an exception to other kinds of credentials where it's possible to run _authenticate and friends.

This is a good point. I haven't thought about it before. It's worth a discussion. I raised a JIRA issue which we can evaluate once the main tasks are done.

> I don't think we covered secondary authentication tests

You are right. I didn't have direct tests for it. It's now added. Thanks!

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java [hunk 6]
```diff
@@ -2159,6 +2163,38 @@ public void testCreateCrossClusterApiKeyMinVersionConstraint() {
...
     }
 
+    public void testAuthenticationFailureWithApiKeyTypeMismatch() throws Exception {
+        final Settings settings = Settings.builder().put(XPackSettings.API_KEY_SERVICE_ENABLED_SETTING.getKey(), true).build();
+        final ApiKeyService service = spy(createApiKeyService(settings));
+
+        final String id = randomAlphaOfLength(12);
+        final String key = randomAlphaOfLength(16);
+        final ApiKey.Type type = randomFrom(ApiKey.Type.values());
+        mockKeyDocument(id, key, new User("hulk", "superuser"), null, true, Duration.ofSeconds(3600), null, type);
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java:2173]: Since the type checking is moved after authentication, we need `invalidated` to be `false` here.

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by ywangd** [comment]
Failure is unrelated I raised https://github.com/elastic/elasticsearch/issues/95970

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by ywangd** [comment]
@elasticmachine run elasticsearch-ci/part-2
